### PR TITLE
Allow using longer ancestor chains in StyleCustomPropertyData

### PR DESCRIPTION
**Pull request by anttijk**
#### 8b7bc22a8c879f1a47bd4966fa30b8a4ce3216ed
<pre>
Allow using longer ancestor chains in StyleCustomPropertyData
<a href="https://bugs.webkit.org/show_bug.cgi?id=261025">https://bugs.webkit.org/show_bug.cgi?id=261025</a>
rdar://114815380

Reviewed by Cameron McCormack.

Currently only one ancestor is allowed. It is typical to have many properties defined on :root.
If we have more definitions further down the tree on multiple levels then all those properties get
copied and we can get large maps duplicated to many elements. This also makes equality comparisons slow.

* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
(WebCore::ComputedStylePropertyMapReadOnly::entries const):
* Source/WebCore/rendering/style/StyleCustomPropertyData.cpp:
(WebCore::StyleCustomPropertyData::StyleCustomPropertyData):

If the parent has more than 8 own properties link to it instead of copying it even
if it already has a parent.
Also limit the maximum ancestor list size to 4.

(WebCore::StyleCustomPropertyData::get const):
(WebCore::StyleCustomPropertyData::set):

Add some asserts.

(WebCore::StyleCustomPropertyData::operator== const):

Shortcut the equality comparison in case the parents are the same.
Use forEach for the slow path.

(WebCore::StyleCustomPropertyData::forEachInternal const):
(WebCore::StyleCustomPropertyData::forEach const):

Traverse all ancestors.
Add IterationStatus return value so iteration can be broken early.
Add templated traversal that avoid Function&lt;&gt; for internal use.

(WebCore::StyleCustomPropertyData::findKeyAtIndex const):
(WebCore::StyleCustomPropertyData::size const):

We now always track the full size so no need to traverse ancestors for it.

* Source/WebCore/rendering/style/StyleCustomPropertyData.h:
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::updateCSSTransitions const):

Canonical link: <a href="https://commits.webkit.org/267595@main">https://commits.webkit.org/267595@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/c54d1db3890c5ac695d2d78730f4ae7037e22f39

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17104 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17429 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17917 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18890 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16004 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17296 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20703 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17571 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18209 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17307 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17663 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14847 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19706 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14903 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15538 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22230 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15902 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15705 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20047 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16290 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13819 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15447 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4083 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19812 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16123 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/rendering/style/StyleCustomPropertyData.cpp [hunk 1]
```diff
@@ -28,104 +28,149 @@
...
+            if (descendant->m_ownValues.contains(key))
+                return true;
+        }
+        return false;
+    };
+
+    auto* propertyData = this;
+    while (true) {
+        for (auto& entry : propertyData->m_ownValues) {
+            if (isOverridenByDescendants(entry.key))
```
Comment (heycam) [inline Source/WebCore/rendering/style/StyleCustomPropertyData.cpp:138]: This is quadratic in the length of the ancestor chain, but since it's short, I guess it's OK in practice, and probably better than building up a `HashSet<AtomString>` of the properties we've seen so far.

File: Source/WebCore/style/Styleable.cpp [hunk 2]
```diff
@@ -692,12 +692,11 @@ void Styleable::updateCSSTransitions(const RenderStyle& currentStyle, const Rend
         auto gatherAnimatableCustomProperties = [&](const StyleCustomPropertyData& customPropertyData) {
             customPropertyData.forEach([&](auto& customPropertyAndValuePair) {
                 auto [customProperty, customPropertyValue] = customPropertyAndValuePair;
-                if (!customPropertyValue)
-                    return;
```
Comment (heycam) [inline Source/WebCore/style/Styleable.cpp:696]: Is this change relevant to this PR, or just a drive by fix?

**Review by heycam — APPROVED** [review]
In the commit message:

> If we more definition further

"If we have more definitions further"

File: Source/WebCore/rendering/style/StyleCustomPropertyData.cpp (hunk 1 above)
Comment (anttijk) [reply Source/WebCore/rendering/style/StyleCustomPropertyData.cpp:138]: Right. With the limit in place the worst case is 4 hash lookup per value. I suspect that is faster than building a new map (which may be large).

File: Source/WebCore/style/Styleable.cpp (hunk 2 above)
Comment (anttijk) [reply Source/WebCore/style/Styleable.cpp:696]: Drive by.
